core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/sync/MixedPackageVersionService.java

```diff
@@ -0,0 +1,226 @@
+            }
+        } else {
+            LOGGER.warn("The validation of the package versions cannot be performed because the CM version is not present in the sync result. {}", syncResult);
+        }
+    }
+
+    private Optional<Image> findTargetImage(Long stackId) {
+        Optional<com.sequenceiq.cloudbreak.cloud.model.Image> targetImage = clusterUpgradeTargetImageService.findTargetImage(stackId);
+        return targetImage.isPresent() ? findImageInCatalog(targetImage.get()) : Optional.empty();
+    }
+
+    private Optional<Image> findImageInCatalog(com.sequenceiq.cloudbreak.cloud.model.Image image) {
+        try {
+            return Optional.of(imageCatalogService.getImage(image.getImageCatalogUrl(), image.getImageCatalogName(), image.getImageId()).getImage());
+        } catch (CloudbreakImageNotFoundException | CloudbreakImageCatalogException e) {
+            LOGGER.warn("There is no image {} found in catalog {}.", image.getImageId(), image.getImageCatalogUrl());
+            return Optional.empty();
+        }
+    }
+
+    private void examinePackageVersionsWithTargetImage(Long stackId, Image targetImage, String activeCmVersion, Set<ParcelInfo> activeParcels) {
+        LOGGER.debug("Comparing active package versions {} and active CM version {} with target image {}", activeParcels, activeCmVersion, targetImage);
+        Map<String, String> targetProducts = getProductsFromImage(targetImage);
+        if (areAllComponentVersionsAreMatchesWithImage(targetImage, targetProducts, activeCmVersion, activeParcels)) {
+            LOGGER.debug("The parcel versions {} and CM {} are the same as the versions on the target image {}", activeParcels, activeCmVersion, targetImage);
+        } else {
+            Map<String, String> newerComponents = getComponentsWithNewerVersionThenTheTarget(targetProducts, targetImage, activeParcels, activeCmVersion);
+            String activeParcelsMessage = createActiveParcelsMessage(activeParcels);
+            if (newerComponents.isEmpty()) {
+                sendNotification(stackId, STACK_CM_MIXED_PACKAGE_VERSIONS_FAILED,
+                        List.of(activeCmVersion, activeParcelsMessage, createSuggestedVersionsMessage(targetImage, activeParcels)));
+            } else {
+                sendNotification(stackId, STACK_CM_MIXED_PACKAGE_VERSIONS_NEWER_FAILED, List.of(activeCmVersion, activeParcelsMessage,
+                        createMessageFromMap(newerComponents),
+                        createMessageFromMap(filterTargetPackageVersionsByNewerPackageVersions(targetProducts, targetImage, newerComponents))));
+            }
+        }
+    }
+
+    private void examinePackageVersionsWithAllCandidateImages(Long stackId, Set<Image> candidateImages, String currentCmVersion, Set<ParcelInfo> activeParcels,
+            String imageCatalogUrl) {
+        LOGGER.debug("Target image not found for cluster therefore comparing package versions with other candidate images.");
```

Comment:
This is also a large private method and I would be nice to have tests for the method. It might make sense to put it to a separate class and create tests for it.